### PR TITLE
[Core] Generalizing condition assign processes in order to be able to assign to any entity

### PR DESCRIPTION
**Pull request by loumalouomega**
As @philbucher mentioned in #3447 some processes are missing for elements, this generalizes the previous ones for conditions to elements and constraints (for scalar)

I will add the python counter part in a different PR

### CODE REVIEW

**Comment by philbucher** [comment]
and again :D
48969479-772df680-efff-11e8-972f-6b3848d3e077

**Comment by loumalouomega** [comment]
I will assign this into @KratosMultiphysics/implementation-committee

**Comment by loumalouomega** [comment]
Ping

**Comment by RiccardoRossi** [comment]
@loumalouomega this touches features used by everyone so we need to be extra careful.
Also i am not positive of doing this changes as we get closer and closer to the release, unless there are important reasons for it.

could u explain the rationale of this change? (please be didactic)

**Comment by loumalouomega** [comment]
1- The current behavior is not broken, the name of the function in python remains the same.
2- This takes the code that was specific for the condition so it can be used in elements too (this can be useful, and I think is a consistent definition)

File: kratos/python/add_processes_to_python.cpp
Comment (philbucher) [inline kratos/python/add_processes_to_python.cpp:372]: You forgot nodes

Comment (loumalouomega) [reply kratos/python/add_processes_to_python.cpp:372]: Nodes already has its own process, and are threaded in a different way as the conditions and elements which have geometries. That's the key, this PR aims to unify and improve without breaking the current code

**Comment by loumalouomega** [comment]
In any case https://github.com/KratosMultiphysics/Kratos/pull/3465 should be checked first

File: kratos/processes/assign_scalar_field_to_entities_process.cpp
Comment (philbucher) [inline kratos/processes/assign_scalar_field_to_entities_process.cpp]:
```suggestion
    const ProcessInfo& r_current_process_info = mrModelPart.GetProcessInfo();
```

File: kratos/processes/assign_scalar_variable_to_entities_process.h
Comment (philbucher) [inline kratos/processes/assign_scalar_variable_to_entities_process.h:218]: I don't see the reason this is needed, it is only used for  array3-component-variables but I don't understand why

Maybe some leftover that should be removed

**Review by philbucher — COMMENTED** [review]
this PR looks good to me and also makes sense in my opinion, thanks for the generalization

Did you grep if the files you rename are included somewhere?

File: kratos/processes/assign_scalar_variable_to_entities_process.h
Comment (loumalouomega) [reply kratos/processes/assign_scalar_variable_to_entities_process.h:218]: I don't' know but the template argument doesn't harm

Comment (philbucher) [reply kratos/processes/assign_scalar_variable_to_entities_process.h:218]: I meant this 
...

File: kratos/processes/assign_scalar_variable_to_entities_process.cpp
Comment (philbucher) [inline kratos/processes/assign_scalar_variable_to_entities_process.cpp:74]: ...
```suggestion
        InternalAssignValue<>(KratosComponents<array_1d_component_type>::Get(mVariableName), mDoubleValue);
```

=> why does the assignment of array3-components have to happen in serial?

File: kratos/processes/assign_scalar_variable_to_entities_process.h
Comment (philbucher) [reply kratos/processes/assign_scalar_variable_to_entities_process.h:218]: answer see below

File: kratos/processes/assign_scalar_variable_to_entities_process.cpp
Comment (loumalouomega) [reply kratos/processes/assign_scalar_variable_to_entities_process.cpp:74]: Huuuuuum, this is quite old, so I don't remember

Comment (loumalouomega) [reply kratos/processes/assign_scalar_variable_to_entities_process.cpp:74]: Apparently without this the tests fail

Comment (philbucher) [reply kratos/processes/assign_scalar_variable_to_entities_process.cpp:74]: Interesting ...

Comment (loumalouomega) [reply kratos/processes/assign_scalar_variable_to_entities_process.cpp:74]: imagen

**Comment by philbucher** [comment]
So from my side this gets a +1 for being merged (after the release)

the strange omp-behavior I pointed out should be investigated but not in this PR

handing over to @KratosMultiphysics/technical-committee

**Comment by maceligueta** [comment]
@loumalouomega , I think it would be great to modify mesh.h and add a method like _GetEntities\<TEntityType\>_ or something similar. Then, maybe all those specializations you are coding in this PR would disappear...

**Comment by loumalouomega** [comment]
> @loumalouomega , I think it would be great to modify mesh.h and add a method like _GetEntities<TEntityType>_ or something similar. Then, maybe all those specializations you are coding in this PR would disappear...

Perhaps. In any case maybe creating a base class for Element and Condition with common methods will allow to simplify a lot of code

**Comment by philbucher** [comment]
Element and Condition have a common baseclass ...

**Comment by loumalouomega** [comment]
> Element and Condition have a common baseclass ...

I meant to create a common class intermediate with the same methods that are defined them in both classes

**Comment by GuillermoCasas** [comment]
> > Element and Condition have a common baseclass ...
> 
> I meant to create a common class intermediate with the same methods that are defined them in both classes

I think this would make a lot of sense. Perhaps you could write a short issue and add it to the to-dos of the @KratosMultiphysics/implementation-committee.

**Comment by philbucher** [comment]
Note that then you have 4 levels of inheritance!
IndexedObject -> GeometricalObject -> IntermediateObject -> Element/Condition

**Comment by loumalouomega** [comment]
> 
> 
> Note that then you have 4 levels of inheritance!
> IndexedObject -> GeometricalObject -> IntermediateObject -> Element/Condition

Of course, but I think the discussion is important

**Comment by philbucher** [comment]
Yes it is important, just adding a point to consider :)

**Comment by jcotela** [comment]
Just a comment, think that whatever is implemented should also work for Nodes, not just Elements/Conditions.

**Comment by loumalouomega** [comment]
> Just a comment, think that whatever is implemented should also work for Nodes, not just Elements/Conditions.

Huum, do you mean a process to set non-historical values on nodes?

**Comment by philbucher** [comment]
@loumalouomega can you please make a separate PR with only the astethic changes, then I will re review and most probably approve, this has been around long enough and I think it is a nice addition
As for the comments of the others, even though this is not perfect it is a significant improvement compared to the current version and prevents lots of code duplication

**Comment by loumalouomega** [comment]
> @loumalouomega can you please make a separate PR with only the astethic changes, then I will re review and most probably approve, this has been around long enough and I think it is a nice addition
> As for the comments of the others, even though this is not perfect it is a significant improvement compared to the current version and prevents lots of code duplication

To this point is a little tricky to separate the aesthetic changes, because mostly the currently existing processes are completely refactored

**Comment by philbucher** [comment]
> > Just a comment, think that whatever is implemented should also work for Nodes, not just Elements/Conditions.
> 
> Huum, do you mean a process to set non-historical values on nodes?

Ok this would be nice, the Process using Get-/SetValue can be used for Nodes,Elems,Conds and Constraints

**Comment by philbucher** [comment]
> > @loumalouomega can you please make a separate PR with only the astethic changes, then I will re review and most probably approve, this has been around long enough and I think it is a nice addition
> > As for the comments of the others, even though this is not perfect it is a significant improvement compared to the current version and prevents lots of code duplication
> 
> To this point is a little tricky to separate the aesthetic changes, because mostly the currently existing processes are completely refactored

Hm ok

If you do the extension to Nodes (Nonhistorical) then I would rereview and most probably approve, other only god knows when this will go in :)

**Comment by loumalouomega** [comment]
Done with the nodes @philbucher and @jcotela

**Comment by philbucher** [comment]
codacy has some (minor) suggestions

**Comment by loumalouomega** [comment]
> codacy has some (minor) suggestions

Done

File: kratos/python_scripts/assign_vector_variable_to_entities_process.py
Comment (philbucher) [inline kratos/python_scripts/assign_vector_variable_to_entities_process.py:69]: please use sth like `for i_dir in range(3):` for this and refactor, this would avoid lots of copy pasting, same in the other processes
=> next PR

File: kratos/python_scripts/assign_vector_variable_to_elements_process.py
Comment (philbucher) [inline kratos/python_scripts/assign_vector_variable_to_elements_process.py:10]:
```suggestion
from KratosMultiphysics import assign_vector_variable_to_entities_process
```
=> next PR

**Review by philbucher — APPROVED** [review]
two more comments, but you can make a separate PR after this one is merged.

Lets hope nothing breaks

nice cleanup :+1:

File: kratos/python_scripts/assign_vector_variable_to_entities_process.py
Comment (philbucher) [reply kratos/python_scripts/assign_vector_variable_to_entities_process.py:69]: this is is almost every process

**Comment by loumalouomega** [comment]
Thanks

**Pull request merged**
